Add Coalesced ID filter support to organizations API

begedin opened the pull request:
Closes #29 

I used logic similar to the one we have in `SkillController`. I also extracted `coalesce_id_string` into `CodeCorps.ControllerHelpers` using the example of `CodeCorps.ModelHelpers`

JoshSmith commented:
@begedin I think admins actually have same abilities as the owner, but the owner can do things like transfer the project to another user/organization. Teams or projects will have a limited number of admins. Not going to be something that happens often, but will be good to spread the work around for overwhelmed project owners.

JoshSmith commented:
I think a DELETE endpoint makes sense. Owners and admins both, as in https://github.com/code-corps/code-corps-api/pull/70#issuecomment-161038242.

JoshSmith commented:
@scarfdestroyer I would do things in this order:

- [ ] Specs for `ContributorPolicy`
- [ ] Add a `DELETE` endpoint that's authed to owners and admins
- [ ] Change admins to have same abilities as owner (for now)
- [ ] Create an issue for us to make it so only admins can transfer ownership of a project to another user/organization

JoshSmith commented:
@begedin this is in progress. I would work on other things as you can without getting held up by this.

JoshSmith commented:
Still in progress, but can be expanded on by whomever wants to take it on.

begedin commented on app/policies/contributor_policy.rb:
```diff
@@ -0,0 +1,58 @@
+
+    return true if @contributor.status == "owner" && current_user_is_owner_of_project?
+    return true if current_user_is_at_least_admin_on_project?
+    return true if @contributor.status == "pending"
```
Are we sure this is the way we want to have it?

The way I see it, any user should be allowed to apply to a project. This is just access rights we're defining here.

Of course, a user should not be able to apply if an application already exists, but that's handled by validation (validating uniqueness of a user scoped to a project) and should return a validation error, not an access right error.

In this definition
* Line 14 means current user will not be able to create a new application unless they already have an application made. There's also a bug here. Need to call `contributor_for_user`, not the `@contributor_for_user`. This way, `@contributor_for_user` will never be set.
* Line 16 means current user can create an owner contributor record if they are also the owner. That makes sense. It means an owner is allowed to transfer ownership to someone else.
* Line 17 means only an admin or an owner can create a contributor record.
* Line 18 means anyone can create a `pending` contributor, but line 14 ensures that only user who are already contributors can do it.

The way I see it, `create?` should return `@user.present? && @contributor.status == "pending"`. This means any logged in user can apply to be a contributor. Only contributor records with "pending" status should be allowed to be created directly.

begedin commented on app/policies/contributor_policy.rb:
```diff
@@ -0,0 +1,58 @@
+      return true if contributor.status_was == "admin" and contributor.collaborator?
+    end
+
+    return false
```
I like the definition of `update?`. It makes sense to me.

Additionally, if we want to remove a user from being a collaborator, that should go into a delete action, I'm thinking for `delete?` admins can delete "pending" and "collaborator", while owners can also delete "admin".

Finally, for ownership transfer, I don't think that should be done on a `Contributor` level at all. Instead, it should go to the project level.

begedin commented:
@scarfdestroyer @JoshSmith I added a comment on how I think the `create?` policy should work. I think we need to discuss it.

I also added a policy for delete? and specs which cover `delete?`, `update?` and `index?`.

I'm thinking there may be confusion with the `create?` policy and access rights, due to the `admin` property on the user model. I don't think it really relates to the contributor model in any way, or am I wrong in this?

scarfdestroyer commented:
The Project model didn't have the `has_many :contributors` so I added that, the counter cache and the migration for the `contributors_count` to store the cache. Also I originally did this on another branch but then realized I branched it from this one so it had all of those failing tests but develop doesn't have the contributor model at all so I closed/deleted it and moved it here.

begedin committed:
Add Coalesced ID filter support to organizations API
